tests/unit/generator_test.py

```diff
@@ -2533,3 +2554,222 @@ def test_session_with_only_null_state(tmp_path: Path) -> None:
+            with RFLX_Types;
+
+            package P with
+              SPARK_Mode
+            is
+
+               type T_Enum is (E1) with
+                 Size =>
+                   63;
+               for T_Enum use (E1 => 1);
+
+               type T (Known : Boolean := False) is
+                  record
+                     case Known is
+                        when True =>
+                           Enum : T_Enum;
+                        when False =>
+                           Raw : RFLX_Types.Base_Integer;
+                     end case;
+                  end record;
+
+               function Valid_T (Unused_Val : RFLX_Types.Base_Integer) return Boolean is
+                 (True);
+
+               function Valid_T (Val : T) return Boolean is
+                 ((if Val.Known then True else Valid_T (Val.Raw) and Val.Raw not in 1));
+
+               function To_Base_Integer (Enum : P.T_Enum) return RFLX_Types.Base_Integer is
+                 ((case Enum is
+                      when E1 =>
+                         1));
+
+               function To_Actual (Enum : T_Enum) return P.T is
+                 ((True, Enum));
+
+               function To_Actual (Val : RFLX_Types.Base_Integer) return P.T is
+                 ((case Val is
+                      when 1 =>
+                         (True, E1),
+                      when others =>
+                         (False, Val)))
+                with
+                 Pre =>
+                   Valid_T (Val);
+
+               function To_Base_Integer (Val : P.T) return RFLX_Types.Base_Integer is
+                 ((if Val.Known then To_Base_Integer (Val.Enum) else Val.Raw));
+
+            end P;
+            """,
+        ),
+    ],
+)
+def test_generate_enumeration_base_type_use(
+    always_valid: bool, expected: str, tmp_path: Path
+) -> None:
+    types = [
+        mty.Enumeration(
+            "P::T",
+            literals=[("E1", expr.Number(1))],
+            size=expr.Number(constants.MAX_SCALAR_SIZE),
+            always_valid=always_valid,
+        )
+    ]
+    Generator(reproducible=True).generate(
+        Model(types=types), Integration(), tmp_path, library_files=False, top_level_package=False
+    )
+    assert (tmp_path / "p.ads").exists()
+    assert (tmp_path / "p.ads").read_text() == textwrap.dedent(expected)
+
+
+def test_generate_optimized_message() -> None:
+    message = Message(
+        "P::Message",
+        [
+            Link(INITIAL, Field("Length")),
+            Link(
+                Field("Length"),
+                Field("Data"),
+                size=expr.Add(expr.Size(expr.Variable("Length")), expr.Number(8)),
+            ),
+            Link(
+                Field("Data"),
+                FINAL,
+            ),
+        ],
+        {
+            Field("Length"): models.UNIVERSAL_LENGTH,
+            Field("Data"): mty.OPAQUE,
+        },
+    )
+    structure = create_structure("", message.proven())
```

Comment:
`message.proven()` does unnecessarily rerun all checks which are already done in the constructor of `Message`.

Reply:
Removed.